### PR TITLE
Type automation fixed (closes #1054)

### LINKED ISSUE #1054: Incorrect typing with enabled 'replace' option
### Are you requesting a feature or reporting a bug?

Bug

### What is the current behavior?

TestCafe set empty string as a value and emulate 'input' event for target element before typing.

### What is the expected behavior?

Some frameworks expect, that target has first symbol of typing text at first input event. Same situation is without testcafe - there is no empty string value, when we select all text and try to type, only first symbol.

### How would you reproduce the current behavior (if this is a bug)?

Run test

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

page markup:
```html
<textarea id="textarea">Initial value</textarea>
<script>
    window.storedValue = '';

    var textarea = document.querySelector('#textarea');

    var storeValue = function () {
        window.storedValue = textarea.value;

        textarea.removeEventListener('input', storeValue);
    };

    textarea.addEventListener('input', storeValue);
</script>
```

Test code

```js
test('Type text with replace', async t => {
    await t
        .typeText('#textarea', 'text', { replace: true })
        .expect(Selector('#textarea').value).eql('text');

    const getFirstValue = ClientFunction(()=> window.storedValue);

    await t.expect(await getFirstValue()).eql('t');
});


```

### Specify your

* operating system: WIN10 x64
* testcafe version: 0.11.1
* node.js version: 6.9.1

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @AlexanderMoskovkin, @helen-dikareva,  @lexkazakov

### CODE REVIEW

File: src/client/automation/playback/type/index.js [hunk 1]
```diff
@@ -137,11 +137,8 @@ export default class TypeAutomation {
         if (!this.replace)
             return;
 
-        if (domUtils.isTextEditableElementAndEditingAllowed(this.element)) {
+        if (domUtils.isTextEditableElementAndEditingAllowed(this.element))
```
Comment (AlexanderMoskovkin) [inline src/client/automation/playback/type/index.js]: we need some refactoring there. The function is called `_clearText` but it's not truth with your changes (we've discussed it in private).
So I think we should do the following:
```js
_calculateTargetElement () {
        var activeElement     = domUtils.getActiveElement();
        var isContentEditable = domUtils.isContentEditableElement(this.element);

        if (isContentEditable) {
            if (activeElement !== contentEditable.findContentEditableParent(this.element)) {
                this.eventState.skipType = true;

                return;
            }
        }
        else if (activeElement !== this.element) {
            this.eventState.skipType = true;

            return;
        }

        this.element = isContentEditable ? this.element : activeElement;
    }

    _type () {
        if (this.eventState.skipType)
            return Promise.resolve();

        if (this.replace) {
            if (domUtils.isTextEditableElementAndEditingAllowed(this.element))
                textSelection.select(this.element);
            else if (isContentEditable)
                textSelection.deleteSelectionContents(this.element, true);
        }         
        ...
    }
```

File: test/functional/fixtures/regression/gh-1054/test.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+describe('[Regression](GH-1054)', function () {
+    describe('Target element should have the first symbol of typing text when the first input event raised', function () {
+        it("Typing in the input element with 'replace' option", function () {
+            return runTests('testcafe-fixtures/index.test.js', 'Type text in the textarea');
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/regression/gh-1054/test.js]: ` it("Typing in the input element with 'replace' option", ...`
`...  'Type text in the textarea');`

??

File: test/functional/fixtures/regression/gh-1054/pages/index.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+<script>
+    window.storedValue = '';
+
+    var textarea = document.querySelector('input');
+    var div      = document.querySelector('div');
+
+    var storeValue = function (event) {
+        var target = event.target;
+
+        window.storedValue = target.value ||target.innerText;
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/regression/gh-1054/pages/index.html]: missed space symbol: `||target.innerText;`

**Comment by georgiy-abbasov** [comment]
FPR

File: test/functional/fixtures/regression/gh-1054/test.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+describe('[Regression](GH-1054)', function () {
+    describe('Target element should have the first symbol of typing text when the first input event raised', function () {
```
Comment (lexkazakov) [inline test/functional/fixtures/regression/gh-1054/test.js]: Target element should contain the first symbol of text input when the input event raised at the first time.

File: test/functional/fixtures/regression/gh-1054/testcafe-fixtures/index.test.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { ClientFunction, Selector } from 'testcafe';
+
+fixture `Check the target value when the first input event raised`
```
Comment (lexkazakov) [inline test/functional/fixtures/regression/gh-1054/testcafe-fixtures/index.test.js]: Check the target element value when the first input event raised

**Pull request merged**